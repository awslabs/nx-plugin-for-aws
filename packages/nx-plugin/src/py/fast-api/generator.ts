/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { FastApiProjectGeneratorSchema } from './schema';
import { UVProvider } from '@nxlv/python/src/provider/uv/provider';
import { Logger } from '@nxlv/python/src/executors/utils/logger';
import pyProjectGenerator, { getPyProjectDetails } from '../project/generator';
import { parse, stringify } from '@iarna/toml';
import { UVPyprojectToml } from '@nxlv/python/src/provider/uv/types';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  sharedConstructsGenerator,
} from '../../utils/shared-constructs';
import { toClassName, toKebabCase, toSnakeCase } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
import { addHttpApi } from '../../utils/http-api';
import { sortProjectTargets } from '../../utils/nx';

/**
 * Generates a Python FastAPI project
 */
export const fastApiProjectGenerator = async (
  tree: Tree,
  schema: FastApiProjectGeneratorSchema,
): Promise<GeneratorCallback> => {
  await sharedConstructsGenerator(tree);

  const { dir, normalizedName, normalizedModuleName, fullyQualifiedName } = getPyProjectDetails(
    tree,
    {
      name: schema.name,
      directory: schema.directory,
    },
  );
  const apiNameSnakeCase = toSnakeCase(schema.name);
  const apiNameKebabCase = toKebabCase(schema.name);
  const apiNameClassName = toClassName(schema.name);
  const enhancedOptions = {
    ...schema,
    dir,
    apiNameClassName,
    apiNameKebabCase,
    apiNameSnakeCase,
  };

  await pyProjectGenerator(tree, {
    name: normalizedName,
    directory: schema.directory,
    moduleName: normalizedModuleName,
    projectType: 'application',
  });

  const projectConfig = readProjectConfiguration(tree, fullyQualifiedName);

  projectConfig.targets.bundle = {
    cache: true,
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${dir}/bundle`],
    options: {
      commands: [
        `uv export --frozen --no-dev --no-editable --project ${normalizedName} -o dist/${dir}/bundle/requirements.txt`,
        `uv pip install --no-installer-metadata --no-compile-bytecode --python-platform x86_64-manylinux2014 --python \`uv python pin\` --target dist/${dir}/bundle -r dist/${dir}/bundle/requirements.txt`
      ],
      parallel: false,
    },
    dependsOn: ['compile'],
  };
  projectConfig.targets.build.dependsOn = [
    ...(projectConfig.targets.build.dependsOn ?? []),
    'bundle',
  ];
  projectConfig.targets.serve = {
    executor: '@nxlv/python:run-commands',
    options: {
      command: `uv run fastapi dev ${normalizedName}/main.py`,
      cwd: dir,
    },
  };

  projectConfig.targets = sortProjectTargets(projectConfig.targets);
  updateProjectConfiguration(tree, normalizedName, projectConfig);

  [
    joinPathFragments(dir, normalizedModuleName ?? normalizedName, 'hello.py'),
    joinPathFragments(dir, 'tests', 'test_hello.py'),
  ].forEach((f) => tree.delete(f));

  generateFiles(
    tree, // the virtual file system
    joinPathFragments(__dirname, 'files', 'app'), // path to the file templates
    dir, // destination path of the files
    {
      name: normalizedName,
    },
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );

  if (
    !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'http-apis',
        `${apiNameKebabCase}.ts`,
      ),
    )
  ) {
    generateFiles(
      tree,
      joinPathFragments(
        __dirname,
        'files',
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
      ),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
      enhancedOptions,
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );

    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'index.ts',
      ),
      './http-apis/index.js',
    );
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'http-apis',
        'index.ts',
      ),
      `./${apiNameKebabCase}.js`,
    );
  }

  addHttpApi(tree, apiNameClassName);

  updateJson(
      tree,
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
      (config: ProjectConfiguration) => {
        if (!config.targets) {
          config.targets = {};
        }
        if (!config.targets.build) {
          config.targets.build = {};
        }
        config.targets.build.dependsOn = [
          ...(config.targets.build.dependsOn ?? []),
          `${fullyQualifiedName}:build`,
        ];
        return config;
      },
    );

  const projectToml = parse(
    tree.read(joinPathFragments(dir, 'pyproject.toml'), 'utf8'),
  ) as UVPyprojectToml;
  projectToml.project.dependencies = ['fastapi', 'mangum'].concat(
    projectToml.project?.dependencies || [],
  );
  projectToml['dependency-groups'] = { dev: ['fastapi[standard]>=0.115'] };
  tree.write(joinPathFragments(dir, 'pyproject.toml'), stringify(projectToml));

  await formatFilesInSubtree(tree);

  return async () => {
    await new UVProvider(tree.root, new Logger(), tree).install();
    installPackagesTask(tree);
  };
};
export default fastApiProjectGenerator;
