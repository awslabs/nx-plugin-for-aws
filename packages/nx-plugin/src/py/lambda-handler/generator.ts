/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { LambdaHandlerProjectGeneratorSchema } from './schema';
import { UVProvider } from '@nxlv/python/src/provider/uv/provider';
import { Logger } from '@nxlv/python/src/executors/utils/logger';
import pyProjectGenerator, { getPyProjectDetails } from '../project/generator';
import { parse, stringify } from '@iarna/toml';
import { UVPyprojectToml } from '@nxlv/python/src/provider/uv/types';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { toClassName, toKebabCase, toSnakeCase } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
import { addLambdaHandler } from '../../utils/lambda-handler';
import { sortObjectKeys } from '../../utils/object';
import { NxGeneratorInfo, getGeneratorInfo } from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';

export const LAMBDA_HANDLER_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

console.log('LAMBDA_HANDLER_GENERATOR_INFO', LAMBDA_HANDLER_GENERATOR_INFO);

/**
 * Generates a Python Lambda Handler project
 */
export const lambdaHandlerProjectGenerator = async (
  tree: Tree,
  schema: LambdaHandlerProjectGeneratorSchema,
): Promise<GeneratorCallback> => {
  await sharedConstructsGenerator(tree);

  const { dir, normalizedName, normalizedModuleName, fullyQualifiedName } =
    getPyProjectDetails(tree, {
      name: schema.name,
      directory: schema.directory,
    });

  const lambdaHandlerClassName = toClassName(schema.name);
  const lambdaHandlerKebabCase = toKebabCase(schema.name);
  const lambdaHandlerSnakeCase = toSnakeCase(schema.name);
  const enhancedOptions = {
    ...schema,
    dir,
    lambdaHandlerClassName,
    lambdaHandlerKebabCase,
    lambdaHandlerSnakeCase,
  };

  await pyProjectGenerator(tree, {
    name: normalizedName,
    directory: schema.directory,
    moduleName: normalizedModuleName,
    projectType: 'application',
  });

  const projectConfig = readProjectConfiguration(tree, fullyQualifiedName);

  projectConfig.targets.bundle = {
    cache: true,
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${dir}/bundle`],
    options: {
      commands: [
        `uv export --frozen --no-dev --no-editable --project ${normalizedName} -o dist/${dir}/bundle/requirements.txt`,
        `uv pip install -n --no-installer-metadata --no-compile-bytecode --python-platform x86_64-manylinux2014 --target dist/${dir}/bundle -r dist/${dir}/bundle/requirements.txt`,
      ],
      parallel: false,
    },
    dependsOn: ['compile'],
  };
  projectConfig.targets.build.dependsOn = [
    ...(projectConfig.targets.build.dependsOn ?? []),
    'bundle',
  ];

  projectConfig.metadata = {
    handlerName: schema.name,
  } as any;

  projectConfig.targets = sortObjectKeys(projectConfig.targets);
  updateProjectConfiguration(tree, normalizedName, projectConfig);

  [
    joinPathFragments(dir, normalizedModuleName ?? normalizedName, 'hello.py'),
    joinPathFragments(dir, 'tests', 'test_hello.py'),
  ].forEach((f) => tree.delete(f));

  generateFiles(
    tree, // the virtual file system
    joinPathFragments(__dirname, 'files', 'app'), // path to the file templates
    dir, // destination path of the files
    {
      name: normalizedName,
      lambdaHandlerClassName,
    },
    {
      overwriteStrategy: OverwriteStrategy.Overwrite,
    },
  );

  if (
    !tree.exists(
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'lambda-handlers',
        `${lambdaHandlerKebabCase}.ts`,
      ),
    )
  ) {
    generateFiles(
      tree,
      joinPathFragments(
        __dirname,
        'files',
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
      ),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
      enhancedOptions,
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );

    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'index.ts',
      ),
      './lambda-handlers/index.js',
    );
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'app',
        'lambda-handlers',
        'index.ts',
      ),
      `./${lambdaHandlerKebabCase}.js`,
    );
  }

  addLambdaHandler(tree, lambdaHandlerKebabCase);

  updateJson(
    tree,
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    (config: ProjectConfiguration) => {
      if (!config.targets) {
        config.targets = {};
      }
      if (!config.targets.build) {
        config.targets.build = {};
      }
      config.targets.build.dependsOn = [
        ...(config.targets.build.dependsOn ?? []),
        `${fullyQualifiedName}:build`,
      ];
      return config;
    },
  );

  const projectToml = parse(
    tree.read(joinPathFragments(dir, 'pyproject.toml'), 'utf8'),
  ) as UVPyprojectToml;
  projectToml.project.dependencies = [
    'aws-lambda-powertools',
    'aws-lambda-powertools[tracer]',
  ].concat(projectToml.project?.dependencies || []);
  tree.write(joinPathFragments(dir, 'pyproject.toml'), stringify(projectToml));

  await addGeneratorMetricsIfApplicable(tree, [LAMBDA_HANDLER_GENERATOR_INFO]);

  await formatFilesInSubtree(tree);

  return async () => {
    await new UVProvider(tree.root, new Logger(), tree).install();
    installPackagesTask(tree);
  };
};
export default lambdaHandlerProjectGenerator;
