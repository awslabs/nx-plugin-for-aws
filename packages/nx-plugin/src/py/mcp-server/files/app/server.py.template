import os
from mcp.server.fastmcp import FastMCP

mcp = FastMCP(
    name="<%- mcpServerNameClassName %>",
    host="0.0.0.0",
    port=int(os.environ.get('PORT', 8000)),
    stateless_http=True,
)

@mcp.tool(description="Adds two numbers")
def add(a: int, b: int) -> int:
    """Add two numbers"""
    return a + b


@mcp.resource("example://context", description="Sample Guidance")
def sample_guidance() -> str:
    return """## Sample Guidance

This is some guidance in markdown format!"""
