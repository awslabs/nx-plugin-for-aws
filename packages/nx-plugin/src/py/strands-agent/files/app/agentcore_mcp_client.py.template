import hashlib
from collections.abc import Generator
from typing import Any

import httpx
from botocore.auth import SigV4Auth
from botocore.awsrequest import AWSRequest
from mcp.client.streamable_http import streamablehttp_client
from strands.tools.mcp.mcp_client import MCPClient


class SigV4HTTPXAuth(httpx.Auth):
    """HTTPX Auth class that signs requests with AWS SigV4."""

    def __init__(self, credentials: Any, region: str):
        self.credentials = credentials
        self.service = "bedrock-agentcore"
        self.region = region
        self.signer = SigV4Auth(credentials, self.service, region)

    def auth_flow(
        self, request: httpx.Request
    ) -> Generator[httpx.Request, httpx.Response, None]:
        headers = dict(request.headers)

        headers.pop("connection", None)
        headers["x-amz-content-sha256"] = hashlib.sha256(
            request.content if request.content else b""
        ).hexdigest()

        aws_request = AWSRequest(
            method=request.method,
            url=str(request.url),
            data=request.content,
            headers=headers,
        )

        self.signer.add_auth(aws_request)

        request.headers.clear()
        request.headers.update(dict(aws_request.headers))

        yield request


class AgentCoreMCPClient:
    """Factory for clients to call MCP servers hosted on Bedrock AgentCore Runtime"""

    @staticmethod
    def _create(
        agent_runtime_arn: str,
        region: str,
        session_id: str,
        headers: dict = None,
        auth_handler: httpx.Auth = None,
    ):
        # Build the URL
        encoded_arn = agent_runtime_arn.replace(":", "%3A").replace("/", "%2F")
        url = f"https://bedrock-agentcore.{region}.amazonaws.com/runtimes/{encoded_arn}/invocations?qualifier=DEFAULT"

        # Create and return the MCP client
        return MCPClient(
            lambda: streamablehttp_client(
                url,
                auth=auth_handler,
                timeout=120,
                terminate_on_close=False,
                headers={
                    "X-Amzn-Bedrock-AgentCore-Runtime-Session-Id": session_id,
                    **(headers if headers is not None else {}),
                },
            )
        )

    @staticmethod
    def with_iam_auth(
        agent_runtime_arn: str, credentials: Any, region: str, session_id: str
    ) -> MCPClient:
        """Create an MCP client with IAM (SigV4) authentication."""
        return AgentCoreMCPClient._create(
            agent_runtime_arn=agent_runtime_arn,
            region=region,
            session_id=session_id,
            auth_handler=SigV4HTTPXAuth(credentials, region),
        )

    @staticmethod
    def with_jwt_auth(
        agent_runtime_arn: str, access_token: str, region: str, session_id: str
    ) -> MCPClient:
        """Create an MCP client with JWT authentication."""
        return AgentCoreMCPClient._create(
            agent_runtime_arn=agent_runtime_arn,
            region=region,
            session_id=session_id,
            headers={
                "Authorization": f"Bearer {access_token}",
            },
        )
