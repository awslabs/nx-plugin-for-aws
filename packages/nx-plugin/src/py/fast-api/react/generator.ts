/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  installPackagesTask,
  joinPathFragments,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
  updateProjectConfiguration,
} from '@nx/devkit';
import { FastApiReactGeneratorSchema } from './schema';
import { runtimeConfigGenerator } from '../../../cloudscape-website/runtime-config/generator';
import snakeCase from 'lodash.snakecase';
import * as path from 'path';
import kebabCase from 'lodash.kebabcase';
import { sortObjectKeys } from '../../../utils/nx';
import { toClassName } from '../../../utils/names';
import { formatFilesInSubtree } from '../../../utils/format';
import { withVersions } from '../../../utils/versions';
import { updateGitIgnore } from '../../../utils/git';

export const fastApiReactGenerator = async (
  tree: Tree,
  options: FastApiReactGeneratorSchema,
) => {
  const frontendProjectConfig = readProjectConfiguration(
    tree,
    options.frontendProjectName,
  );
  const fastApiProjectConfig = readProjectConfiguration(
    tree,
    options.fastApiProjectName,
  );
  const moduleName = getFastApiModuleName(fastApiProjectConfig);

  // Add OpenAPI spec generation script to FastAPI spec (if it does not exist already)
  generateFiles(
    tree,
    path.join(__dirname, 'files/fast-api'),
    fastApiProjectConfig.root,
    {
      moduleName,
    },
  );

  // Instrument the script as part of the fastapi project build
  const fastApiOpenApiDist = joinPathFragments(
    'dist',
    fastApiProjectConfig.root,
    'openapi',
  );
  const specPath = joinPathFragments(fastApiOpenApiDist, 'openapi.json');
  updateProjectConfiguration(tree, options.fastApiProjectName, {
    ...fastApiProjectConfig,
    targets: sortObjectKeys({
      ...fastApiProjectConfig.targets,
      build: {
        ...fastApiProjectConfig.targets?.build,
        dependsOn: [
          ...(fastApiProjectConfig.targets?.build?.dependsOn ?? []).filter(
            (t) => t !== 'openapi',
          ),
          'openapi',
        ],
      },
      openapi: {
        cache: true,
        executor: 'nx:run-commands',
        outputs: [joinPathFragments('{workspaceRoot}', fastApiOpenApiDist)],
        options: {
          commands: [
            `uv run python ${joinPathFragments(fastApiProjectConfig.root, 'scripts', 'generate_open_api.py')} "${specPath}"`,
          ],
        },
      },
    }),
  });

  const apiName = (fastApiProjectConfig.metadata as any)?.apiName;
  const clientGenTarget = `generate:${kebabCase(apiName)}-client`;

  const generatedClientDir = joinPathFragments('generated', kebabCase(apiName));
  const generatedClientDirFromRoot = joinPathFragments(
    frontendProjectConfig.sourceRoot,
    generatedClientDir,
  );

  // Add TypeScript client generation to Frontend project.json
  updateProjectConfiguration(tree, options.frontendProjectName, {
    ...frontendProjectConfig,
    targets: sortObjectKeys({
      ...frontendProjectConfig.targets,
      // Generate should run before compile as the client is created as part of the website src
      compile: {
        ...frontendProjectConfig.targets?.compile,
        dependsOn: [
          ...(frontendProjectConfig.targets?.compile?.dependsOn ?? []).filter(
            (t) => t !== clientGenTarget,
          ),
          clientGenTarget,
        ],
      },
      [clientGenTarget]: {
        cache: true,
        executor: 'nx:run-commands',
        inputs: [
          {
            dependentTasksOutputFiles: '**/*.json',
          },
        ],
        outputs: [
          joinPathFragments('{workspaceRoot}', generatedClientDirFromRoot),
        ],
        options: {
          commands: [
            `nx g @aws/nx-plugin:open-api#ts-client --openApiSpecPath="${specPath}" --outputPath="${generatedClientDirFromRoot}" --no-interactive`,
          ],
        },
        dependsOn: [`${options.fastApiProjectName}:openapi`],
      },
    }),
  });

  const relativeSrcDir = frontendProjectConfig.sourceRoot.slice(
    frontendProjectConfig.root.length + 1,
  );

  // Ignore the generated client by default
  // Users can safely remove the entry from the .gitignore if they prefer to check it in
  updateGitIgnore(tree, frontendProjectConfig.root, (patterns) => [
    ...patterns,
    joinPathFragments(relativeSrcDir, generatedClientDir),
  ]);

  // Ensure that the frontend has runtime config as we'll use the url for creating the client
  await runtimeConfigGenerator(tree, {
    project: options.frontendProjectName,
  });

  // Add sigv4 fetch
  if (options.auth === 'IAM') {
    generateFiles(
      tree,
      joinPathFragments(__dirname, '../../../utils/files/website/hooks/sigv4'),
      joinPathFragments(frontendProjectConfig.sourceRoot, 'hooks'),
      {},
    );
  }

  // Add a hook to instantiate the client
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'website'),
    frontendProjectConfig.sourceRoot,
    {
      auth: options.auth,
      apiName,
      apiNameClassName: toClassName(apiName),
      generatedClientDir,
    },
  );

  addDependenciesToPackageJson(
    tree,
    withVersions([
      ...((options.auth === 'IAM'
        ? [
            'oidc-client-ts',
            'react-oidc-context',
            '@aws-sdk/client-cognito-identity',
            '@aws-sdk/credential-provider-cognito-identity',
            'aws4fetch',
          ]
        : []) as any),
    ]),
    withVersions(['@smithy/types']),
  );

  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
};

const getFastApiModuleName = (projectConfig: ProjectConfiguration): string => {
  if (projectConfig.sourceRoot) {
    const sourceRootParts = projectConfig.sourceRoot.split('/');
    return sourceRootParts[sourceRootParts.length - 1];
  }
  const apiName = (projectConfig.metadata as any)?.apiName;
  if (apiName) {
    return snakeCase(apiName);
  }
  new Error(`Could not determine sourceRoot for project ${projectConfig.name}`);
};

export default fastApiReactGenerator;
