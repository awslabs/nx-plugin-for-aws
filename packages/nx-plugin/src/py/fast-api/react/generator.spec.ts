/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { fastApiReactGenerator } from './generator';
import { createTreeUsingTsSolutionSetup } from '../../../utils/test';
import { query } from '../../../utils/ast';

describe('fastapi react generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
    // Mock frontend project configuration
    tree.write(
      'apps/frontend/project.json',
      JSON.stringify({
        name: 'frontend',
        root: 'apps/frontend',
        sourceRoot: 'apps/frontend/src',
      }),
    );
    // Mock FastAPI project configuration
    tree.write(
      'apps/backend/project.json',
      JSON.stringify({
        name: 'backend',
        root: 'apps/backend',
        sourceRoot: 'apps/backend/src',
        metadata: {
          apiName: 'TestApi',
        },
      }),
    );
    // Mock main.tsx file
    tree.write(
      'apps/frontend/src/main.tsx',
      `
import { App } from './app';
import { RouterProvider } from '@tanstack/react-router';

export function Main() {
  return <RouterProvider router={router} />;
}
`,
    );
  });

  it('should generate OpenAPI spec script', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    // Verify OpenAPI spec generation script was created
    expect(
      tree.exists('apps/backend/scripts/generate_open_api.py'),
    ).toBeTruthy();

    // Create snapshot of generated script
    expect(
      tree.read('apps/backend/scripts/generate_open_api.py', 'utf-8'),
    ).toMatchSnapshot('generate_open_api.py');
  });

  it('should update FastAPI project configuration', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    const projectConfig = JSON.parse(
      tree.read('apps/backend/project.json', 'utf-8'),
    );

    // Verify openapi target was added
    expect(projectConfig.targets.openapi).toBeDefined();
    expect(projectConfig.targets.openapi.executor).toBe('nx:run-commands');
    expect(projectConfig.targets.openapi.outputs).toEqual([
      '{workspaceRoot}/dist/apps/backend/openapi',
    ]);

    // Verify build target was updated to depend on openapi
    expect(projectConfig.targets.build.dependsOn).toContain('openapi');
  });

  it('should update frontend project configuration', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    const projectConfig = JSON.parse(
      tree.read('apps/frontend/project.json', 'utf-8'),
    );

    // Verify client generation target was added
    expect(projectConfig.targets['generate:test-api-client']).toBeDefined();
    expect(projectConfig.targets['generate:test-api-client'].executor).toBe(
      'nx:run-commands',
    );

    expect(
      projectConfig.targets['generate:test-api-client'].options.commands,
    ).toEqual([
      'nx g @aws/nx-plugin:open-api#ts-hooks --openApiSpecPath="dist/apps/backend/openapi/openapi.json" --outputPath="apps/frontend/src/generated/test-api" --no-interactive',
    ]);

    expect(
      projectConfig.targets['generate:test-api-client'].dependsOn,
    ).toContain('backend:openapi');

    // Verify generated client is ignored by default
    expect(tree.exists('apps/frontend/.gitignore'));
    expect(tree.read('apps/frontend/.gitignore', 'utf-8')).toContain(
      'src/generated/test-api',
    );

    // Verify compile target depends on client generation
    expect(projectConfig.targets.compile.dependsOn).toContain(
      'generate:test-api-client',
    );

    // Verify bundle target depends on client generation
    expect(projectConfig.targets.bundle.dependsOn).toContain(
      'generate:test-api-client',
    );
  });

  it('should generate vanilla client hook', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    // Verify hook file was created
    expect(
      tree.exists('apps/frontend/src/hooks/useTestApiClient.tsx'),
    ).toBeTruthy();

    // Create snapshot of generated hook
    expect(
      tree.read('apps/frontend/src/hooks/useTestApiClient.tsx', 'utf-8'),
    ).toMatchSnapshot('useTestApiClient.tsx');
  });

  it('should generate tanstack query options proxy hook', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    // Verify hook file was created
    expect(tree.exists('apps/frontend/src/hooks/useTestApi.tsx')).toBeTruthy();

    // Create snapshot of generated hook
    expect(
      tree.read('apps/frontend/src/hooks/useTestApi.tsx', 'utf-8'),
    ).toMatchSnapshot('useTestApi.tsx');
  });

  it('should generate tanstack query options proxy provider', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    // Verify providers were created
    expect(
      tree.exists('apps/frontend/src/components/QueryClientProvider.tsx'),
    ).toBeTruthy();
    expect(
      tree.exists('apps/frontend/src/components/TestApiProvider.tsx'),
    ).toBeTruthy();

    // Create snapshot of generated provider
    expect(
      tree.read('apps/frontend/src/components/TestApiProvider.tsx', 'utf-8'),
    ).toMatchSnapshot('TestApiProvider.tsx');
  });

  it('should instrument providers in main.tsx', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    expect(
      query(
        tree,
        'apps/frontend/src/main.tsx',
        'JsxOpeningElement[tagName.name="QueryClientProvider"]',
      ),
    ).toHaveLength(1);

    expect(
      query(
        tree,
        'apps/frontend/src/main.tsx',
        'JsxOpeningElement[tagName.name="TestApiProvider"]',
      ),
    ).toHaveLength(1);

    expect(tree.read('apps/frontend/src/main.tsx', 'utf-8')).toMatchSnapshot(
      'main.tsx',
    );
  });

  it('should not duplicate providers in main.tsx', async () => {
    // Generate twice
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'None',
    });

    // Check that there is still only a single QueryClientProvider
    expect(
      query(
        tree,
        'apps/frontend/src/main.tsx',
        'JsxOpeningElement[tagName.name="QueryClientProvider"]',
      ),
    ).toHaveLength(1);

    // Check that there is still only a single TestApiProvider
    expect(
      query(
        tree,
        'apps/frontend/src/main.tsx',
        'JsxOpeningElement[tagName.name="TestApiProvider"]',
      ),
    ).toHaveLength(1);
  });

  it('should handle IAM auth option', async () => {
    await fastApiReactGenerator(tree, {
      frontendProjectName: 'frontend',
      fastApiProjectName: 'backend',
      auth: 'IAM',
    });

    // Verify sigv4 hook was added
    expect(tree.exists('apps/frontend/src/hooks/useSigV4.tsx')).toBeTruthy();

    // Verify IAM-specific dependencies were added
    const packageJson = JSON.parse(tree.read('package.json', 'utf-8'));
    expect(packageJson.dependencies['oidc-client-ts']).toBeDefined();
    expect(packageJson.dependencies['react-oidc-context']).toBeDefined();
    expect(
      packageJson.dependencies['@aws-sdk/client-cognito-identity'],
    ).toBeDefined();
    expect(
      packageJson.dependencies['@aws-sdk/credential-provider-cognito-identity'],
    ).toBeDefined();
    expect(packageJson.dependencies['aws4fetch']).toBeDefined();

    // Create snapshot of generated hook with IAM auth
    expect(
      tree.read('apps/frontend/src/hooks/useTestApiClient.tsx', 'utf-8'),
    ).toMatchSnapshot('useTestApi-IAM.tsx');
  });
});
