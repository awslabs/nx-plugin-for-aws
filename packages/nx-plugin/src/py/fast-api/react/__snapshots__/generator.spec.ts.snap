// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fastapi react generator > should generate OpenAPI spec script > generate_open_api.py 1`] = `
"from src.main import app
import json, os, sys

os.makedirs(os.path.dirname(sys.argv[1]), exist_ok=True)
with open(sys.argv[1], 'w') as f:
  json.dump(app.openapi(), f)
"
`;

exports[`fastapi react generator > should generate client hook > useTestApi.tsx 1`] = `
"import { TestApi } from '../generated/test-api/client.gen';
import { useRuntimeConfig } from './useRuntimeConfig';
import { useMemo } from 'react';

export const useTestApi = (): TestApi => {
  const runtimeConfig = useRuntimeConfig();
  const apiUrl = runtimeConfig.httpApis.TestApi;
  return useMemo(
    () =>
      new TestApi({
        url: apiUrl,
      }),
    [apiUrl],
  );
};
"
`;

exports[`fastapi react generator > should handle IAM auth option > useTestApi-IAM.tsx 1`] = `
"import { TestApi } from '../generated/test-api/client.gen';
import { useSigV4 } from './useSigV4';
import { useRuntimeConfig } from './useRuntimeConfig';
import { useMemo } from 'react';

export const useTestApi = (): TestApi => {
  const runtimeConfig = useRuntimeConfig();
  const apiUrl = runtimeConfig.httpApis.TestApi;
  const sigv4Client = useSigV4();
  return useMemo(
    () =>
      new TestApi({
        url: apiUrl,
        fetch: sigv4Client,
      }),
    [apiUrl, sigv4Client],
  );
};
"
`;
